lib/location/method_gnss.c

```diff
@@ -0,0 +1,341 @@
+			method_gnss_cancel();
+			loc_core_event_cb(&location_result);
+		} else {
+			fix_attemps_remaining--;
+		}
+	}
+}
+
+void method_gnss_timeout_work_fn(struct k_work *item)
+{
+	loc_core_event_cb_timeout();
+
+	method_gnss_cancel();
+}
+
+static void method_gnss_positioning_work_fn(struct k_work *work)
+{
+	int err = 0;
+	int tau;
+	int active_time;
+	enum lte_lc_system_mode mode;
+
+	struct k_work_args *work_data =
+		CONTAINER_OF(work, struct k_work_args, work_item);
+
+	const struct loc_gnss_config gnss_config = work_data->gnss_config;
+
+	lte_lc_system_mode_get(&mode, NULL);
+
+	/* Don't care about PSM if we are in GNSS only mode */
+	if(mode != LTE_LC_SYSTEM_MODE_GPS) {
+		int ret = lte_lc_psm_get(&tau, &active_time);
+		if (ret < 0) {
+			LOG_ERR("Cannot get PSM config: %d. Starting GNSS right away.", ret);
+		} else if ((tau >= 0) & (active_time >= 0)) {
+			LOG_INF("Waiting for LTE to enter PSM: tau %d active_time %d",
+				tau, active_time);
+
+			/* Wait for the PSM to start. If semaphore is reset during the waiting
+			 * period, the position request was canceled. Thus, return without
+			 * doing anything */
+			if (k_sem_take(&entered_psm_mode, K_FOREVER) == -EAGAIN) {
+				return;
+			}
+			k_sem_give(&entered_psm_mode);
+		}
+	}
+
+	/* Configure GNSS to continuous tracking mode */
+	err = nrf_modem_gnss_fix_interval_set(1);
+
+	uint8_t use_case;
+
+	switch (gnss_config.accuracy) {
+	case LOC_ACCURACY_LOW:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START |
+			   NRF_MODEM_GNSS_USE_CASE_LOW_ACCURACY;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOC_ACCURACY_NORMAL:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOC_ACCURACY_HIGH:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		if (!err) {
+			fix_attemps_remaining = gnss_config.num_consecutive_fixes;
+		}
+		break;
+	}
+
+	if (err) {
+		LOG_ERR("Failed to configure GNSS");
+		loc_core_event_cb_error();
+		running = false;
+	}
+
+	err = nrf_modem_gnss_start();
+	if (err) {
+		LOG_ERR("Failed to start GNSS");
+		loc_core_event_cb_error();
+		running = false;
+	}
+
+	LOG_DBG("GNSS started");
+}
+
+void method_gnss_modem_sleep_notif_subscribe(uint32_t threshold_ms)
+{
+	char buf_sub[48];
+	int err;
+
+	snprintk(buf_sub, sizeof(buf_sub), AT_MDM_SLEEP_NOTIF_START,
+		 0, threshold_ms);
+
+	err = at_cmd_write(buf_sub, NULL, 0, NULL);
+	if (err) {
+		LOG_ERR("Cannot subscribe to modem sleep notifications, err %d", err);
+	} else {
+		LOG_DBG("Subscribed to modem sleep notifications");
+	}
+}
+
+int method_gnss_location_get(const struct loc_method_config *config)
+{
+	const struct loc_gnss_config gnss_config = config->gnss;
+
+	LOG_INF("Starting to get a location by using gnss method");
+
+	if (running) {
+		LOG_ERR("Previous operation ongoing.");
+		return -EBUSY;
+	}
+
+	method_gnss_agps_request();
```

Comment:
I guess we'd like to have a way to decide whether to use AGPS data or not in MOSH when testing. For real applications, this is likely not a concern as they decide it with configuration `CONFIG_NRF_CLOUD_AGPS`.
Having a configuration option in `struct loc_gnss_config` is one option but let's think about it.